src/transformers/modeling_dpr.py

```diff
@@ -0,0 +1,551 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch DPR model for Open Domain Question Answering."""
+
+
+import collections
+import logging
+from typing import List, Optional, Tuple
+
+import torch
+from torch import Tensor as T
+from torch import nn
+from torch.serialization import default_restore_location
+
+from .configuration_dpr import DprConfig
+from .file_utils import add_start_docstrings
+from .modeling_bert import BertConfig, BertModel
+from .modeling_utils import PreTrainedModel
+
+
+logger = logging.getLogger(__name__)
+
+DPR_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "dpr-base-uncased",
+]
+
+
+#############
+# BertEncoder
+#############
+
+
+class DprBertEncoder(BertModel):
+    def __init__(self, config, project_dim: int = 0):
+        BertModel.__init__(self, config)
+        assert config.hidden_size > 0, "Encoder hidden_size can't be zero"
+        self.encode_proj = nn.Linear(config.hidden_size, project_dim) if project_dim != 0 else None
+        self.init_weights()
+
+    @classmethod
+    def init_encoder(cls, cfg_name: str, projection_dim: int = 0, dropout: float = 0.0, **kwargs) -> "DprBertEncoder":
+        cfg = BertConfig.from_pretrained(cfg_name)
+        if dropout != 0:
+            cfg.attention_probs_dropout_prob = dropout
+            cfg.hidden_dropout_prob = dropout
+        return cls(cfg, project_dim=projection_dim)
+
+    def forward(self, input_ids: T, token_type_ids: Optional[T], attention_mask: Optional[T]) -> Tuple[T, ...]:
+        if self.config.output_hidden_states:
+            sequence_output, pooled_output, hidden_states = super().forward(
+                input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
+            )
+        else:
+            hidden_states = None
+            sequence_output, pooled_output = super().forward(
+                input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
+            )
+        pooled_output = sequence_output[:, 0, :]
+        if self.encode_proj:
+            pooled_output = self.encode_proj(pooled_output)
+        return sequence_output, pooled_output, hidden_states
+
+    def get_out_size(self) -> int:
+        if self.encode_proj:
+            return self.encode_proj.out_features
+        return self.config.hidden_size
+
+
+###########
+# Reader
+###########
+
+
+class Reader(nn.Module):
+    def __init__(self, encoder: nn.Module, hidden_size):
+        super(Reader, self).__init__()
+        self.encoder = encoder
+        self.qa_outputs = nn.Linear(hidden_size, 2)
+        self.qa_classifier = nn.Linear(hidden_size, 1)
```

Comment:
This looks like it could inherit from `PretrainedModel` instead of `nn.Module`, and would therefore be able to use all methods/helpers from that class.

Reply:
I just did the change thanks. Can you check that the `init_weights` make sense ?